File: backend/src/main/kotlin/com/logvue/data/model/LogEntry.kt
package com.logvue.data.model

import kotlinx.serialization.Serializable

@Serializable
data class LogEntry(
    val id: Int,
    val header: LogHeader,
    val message: String,
    val timestamp: Long
)

@Serializable
data class LogHeader(
    val logLevel: String,
    val pid: Int,
    val tid: Int,
    val applicationId: String? = null,
    val processName: String? = null,
    val tag: String? = null,
    val timestamp: Timestamp
)

@Serializable
data class Timestamp(
    val seconds: Long,
    val nanos: Int
)

@Serializable
data class FilterRequest(
    val fileId: String,
    val levels: List<String> = emptyList(),
    val tagPattern: String = "",
    val tagRegex: Boolean = false,
    val contentFilter: String = "",
    val searchQuery: String? = null,
    val searchCaseSensitive: Boolean = false,
    val searchRegex: Boolean = false,
    val timeFrom: Long? = null,
    val timeTo: Long? = null,
    val hiddenTags: String = "",
    val offset: Int = 0,
    val limit: Int = 200
)

@Serializable
data class FilterResponse(
    val entries: List<LogEntry>,
    val total: Int,
    val hasMore: Boolean,
    val searchHighlightRanges: Map<String, List<List<Int>>> = emptyMap(),
    val levelCounts: Map<String, Int> = emptyMap(),
    val tagCounts: List<TagCount> = emptyList()
)

@Serializable
data class TagCount(
    val tag: String,
    val count: Int
)

File: backend/models/log_entry.go
package models

// Timestamp represents structured time with seconds and nanos
type Timestamp struct {
	Seconds int64 `json:"seconds"`
	Nanos   int   `json:"nanos"`
}

// LogHeader contains metadata for each log entry
type LogHeader struct {
	LogLevel     string    `json:"logLevel"`
	PID          int       `json:"pid"`
	TID          int       `json:"tid"`
	ApplicationID string   `json:"applicationId,omitempty"`
	ProcessName  string   `json:"processName,omitempty"`
	Tag          string   `json:"tag,omitempty"`
	Timestamp    Timestamp `json:"timestamp"`
	ThreadName   string   `json:"threadName,omitempty"`
}

// LogEntry represents a single log entry
type LogEntry struct {
	ID        int       `json:"id"`
	Header    LogHeader `json:"header"`
	Message   string    `json:"message"`
	Timestamp int64     `json:"timestamp"`
}

// TagCount represents tag usage statistics
type TagCount struct {
	Tag   string `json:"tag"`
	Count int    `json:"count"`
}

// FilterRequest is the request object for filtering logs
type FilterRequest struct {
	FileID             string   `json:"fileId"`
	Levels             []string `json:"levels"`
	TagPattern         string   `json:"tagPattern"`
	TagRegex           bool     `json:"tagRegex"`
	ContentFilter      string   `json:"contentFilter"`
	SearchQuery        string   `json:"searchQuery,omitempty"`
	SearchCaseSensitive bool   `json:"searchCaseSensitive"`
	SearchRegex        bool     `json:"searchRegex"`
	TimeFrom           *int64   `json:"timeFrom,omitempty"`
	TimeTo             *int64   `json:"timeTo,omitempty"`
	HiddenTags         string   `json:"hiddenTags"`
	Offset             int      `json:"offset"`
	Limit              int      `json:"limit"`
}

// FilterResponse is the response object from filtering
type FilterResponse struct {
	Entries               []LogEntry                       `json:"entries"`
	Total                 int                             `json:"total"`
	HasMore               bool                            `json:"hasMore"`
	SearchHighlightRanges map[string][][]int              `json:"searchHighlightRanges"`
	LevelCounts           map[string]int                  `json:"levelCounts"`
	TagCounts             []TagCount                      `json:"tagCounts"`
}
